staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -0,0 +1,386 @@
+		}
+
+		rw := NewCustomResponseWriter(w)
+		h.proxyRequestToDestinationAPIServer(r, rw, destServerIP)
+
+	})
+}
+
+func NewCustomResponseWriter(w http.ResponseWriter) *CustomResponseWriter {
+	return &CustomResponseWriter{ResponseWriter: w, statusCode: http.StatusOK}
+}
+
+func (h *peerProxyHandler) findServiceableByServers(gvr schema.GroupVersionResource, localAPIServerId string, reconciler reconcilers.PeerEndpointLeaseReconciler) (serviceableByResponse, error) {
+
+	apiserversi, ok := h.svMap.Load(gvr)
+
+	// no value found for the requested gvr in svMap
+	if !ok || apiserversi == nil {
+		return serviceableByResponse{}, fmt.Errorf("no StorageVersions found for the GVR: %v", gvr)
+	}
+	apiservers := apiserversi.(*sync.Map)
+	response := serviceableByResponse{}
+	var peerServerIPs []string
+	apiservers.Range(func(key, value interface{}) bool {
+		apiserverKey := key.(string)
+		if apiserverKey == localAPIServerId {
+			response.locallyServiceable = true
+			// stop iteration
+			return false
+		}
+
+		ip, err := reconciler.GetEndpoint(apiserverKey)
+		if err != nil {
+			// continue with iteration
+			klog.Errorf("failed to get peer ip from storage lease for server %s", apiserverKey)
+			return true
+		}
+		peerServerIPs = append(peerServerIPs, ip)
+		// continue with iteration
+		return true
+	})
+	response.peerIPs = peerServerIPs
+	return response, nil
+}
+
+func (h *peerProxyHandler) proxyRequestToDestinationAPIServer(req *http.Request, crw *CustomResponseWriter, host string) {
+	user, ok := apirequest.UserFrom(req.Context())
+	if !ok {
+		klog.Errorf("failed to get user info from request")
+		return
+	}
+
+	// write a new location based on the existing request pointed at the target service
+	location := &url.URL{}
+	location.Scheme = "https"
+	location.Host = host
+	location.Path = req.URL.Path
+	location.RawQuery = req.URL.Query().Encode()
+
+	newReq, cancelFn := apiserverproxyutil.NewRequestForProxy(location, req)
+	newReq.Header.Add(PeerProxiedHeader, "true")
+	defer cancelFn()
+
+	// create transport
+	clientConfig := &transport.Config{
+		TLS: transport.TLSConfig{
+			Insecure:   false,
+			CertFile:   h.proxyClientCertFile,
+			KeyFile:    h.proxyClientKeyFile,
+			CAFile:     h.peerCAFile,
+			ServerName: "kubernetes.default.svc",
+		},
+	}
+
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	defer cancelCtx()
+	proxyRoundTripper, transportBuildingError := transport.New(clientConfig)
+	if transportBuildingError != nil {
+		klog.Error(transportBuildingError.Error())
+		return
+	}
+
+	proxyRoundTripper = transport.NewAuthProxyRoundTripper(user.GetName(), user.GetGroups(), nil, proxyRoundTripper)
+
+	handler := proxy.NewUpgradeAwareHandler(location, proxyRoundTripper, true, false, &responder{w: crw, ctx: req.Context()})
+	handler.ServeHTTP(crw, newReq)
+	// Increment the count of proxied requests
+	metrics.IncPeerProxiedRequest(ctx, strconv.Itoa(crw.statusCode))
+}
+
+func (crw *CustomResponseWriter) WriteHeader(code int) {
```

Comment:
Does this need to support the case where WriteHeader is never called? Also, curious whether this could reuse the ResponseWriter wrapper implementation from k8s.io/apiserver/pkg/endpoints/metrics.

Reply:
Yep, reused endpoints/metrics 's ResponseWriter. Thanks again!